File: examples/bugbot/ts/src/validate.ts
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

export type BugbotFinding = {
  rule: string;
  severity: "blocking" | "non-blocking";
  title: string;
  path: string;
  detail: string;
};

const exampleRoot = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../..",
);

export function resolveDiffPath(override?: string): string {
  const trimmed = override?.trim();
  if (trimmed) {
    return path.isAbsolute(trimmed)
      ? trimmed
      : path.resolve(process.cwd(), trimmed);
  }

  return path.join(exampleRoot, "fixtures/pr-idempotency-bug.diff");
}

export function listChangedFiles(diff: string): string[] {
  const files = new Set<string>();

  for (const line of diff.split("\n")) {
    const match = line.match(/^\+\+\+ b\/(.+)$/);
    if (match?.[1] && match[1] !== "/dev/null") {
      files.add(match[1]);
    }
  }

  return [...files];
}

export function addedLinesForFile(diff: string, filePath: string): string[] {
  const lines: string[] = [];
  let currentFile: string | null = null;
  let inHunk = false;

  for (const line of diff.split("\n")) {
    const fileMatch = line.match(/^\+\+\+ b\/(.+)$/);
    if (fileMatch?.[1]) {
      currentFile = fileMatch[1];
      inHunk = false;
      continue;
    }

    if (line.startsWith("@@")) {
      inHunk = currentFile === filePath;
      continue;
    }

    if (!inHunk || currentFile !== filePath) {
      continue;
    }

    if (line.startsWith("+") && !line.startsWith("+++")) {
      lines.push(line.slice(1));
    }
  }

  return lines;
}

export function validateBugbotDiff(diff: string): BugbotFinding[] {
  const findings: BugbotFinding[] = [];
  const changedFiles = listChangedFiles(diff);
  const paymentsChanged = changedFiles.some((file) =>
    file.startsWith("src/payments/"),
  );
  const paymentsTestsChanged = changedFiles.some((file) =>
    file.startsWith("tests/payments/"),
  );

  const checkoutAdds = addedLinesForFile(diff, "src/payments/checkout.ts");
  const checkoutText = checkoutAdds.join("\n");

  if (
    /idempotencyKey:\s*token\b/.test(checkoutText) ||
    /idempotencyKey:\s*[^,}\n]+token/.test(checkoutText)
  ) {
    findings.push({
      rule: "unsafe-idempotency-key",
      severity: "blocking",
      title: "Unsafe idempotency key",
      path: "src/payments/checkout.ts",
      detail:
        "Client token reused as idempotencyKey. Generate a server-side UUID instead.",
    });
  }

  if (
    paymentsChanged &&
    /gateway\.refund|\/refunds/.test(checkoutText) &&
    !/authorize|ownership|role|permission/i.test(checkoutText)
  ) {
    findings.push({
      rule: "refund-missing-authorization",
      severity: "blocking",
      title: "Refund path missing authorization",
      path: "src/payments/checkout.ts",
      detail:
        "New refund path calls gateway.refund without an authorization check.",
    });
  }

  if (paymentsChanged && !paymentsTestsChanged) {
    findings.push({
      rule: "missing-payments-tests",
      severity: "blocking",
      title: "Missing payments tests",
      path: "src/payments/",
      detail:
        "Payment handler files changed without matching tests under tests/payments/.",
    });
  }

  return findings;
}

export async function loadDefaultDiff(): Promise<string> {
  return fs.readFile(resolveDiffPath(), "utf8");
}

export function formatFindings(findings: BugbotFinding[]): string {
  if (findings.length === 0) {
    return "No Bugbot rule violations detected in the fixture diff.";
  }

  return findings
    .map(
      (finding) =>
        `[${finding.severity}] ${finding.title}\n  rule: ${finding.rule}\n  path: ${finding.path}\n  ${finding.detail}`,
    )
    .join("\n\n");
}

File: examples/bugbot/python/validate.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

Severity = Literal["blocking", "non-blocking"]

EXAMPLE_ROOT = Path(__file__).resolve().parents[1]


@dataclass(frozen=True)
class BugbotFinding:
    rule: str
    severity: Severity
    title: str
    path: str
    detail: str


def resolve_diff_path(override: str | None = None) -> Path:
    if override and override.strip():
        candidate = Path(override.strip())
        return candidate if candidate.is_absolute() else Path.cwd() / candidate
    return EXAMPLE_ROOT / "fixtures" / "pr-idempotency-bug.diff"


def list_changed_files(diff: str) -> list[str]:
    files: list[str] = []
    seen: set[str] = set()
    for line in diff.splitlines():
        match = re.match(r"^\+\+\+ b/(.+)$", line)
        if match and match.group(1) != "/dev/null":
            path = match.group(1)
            if path not in seen:
                seen.add(path)
                files.append(path)
    return files


def added_lines_for_file(diff: str, file_path: str) -> list[str]:
    lines: list[str] = []
    current_file: str | None = None
    in_hunk = False
    for line in diff.splitlines():
        file_match = re.match(r"^\+\+\+ b/(.+)$", line)
        if file_match:
            current_file = file_match.group(1)
            in_hunk = False
            continue
        if line.startswith("@@"):
            in_hunk = current_file == file_path
            continue
        if not in_hunk or current_file != file_path:
            continue
        if line.startswith("+") and not line.startswith("+++"):
            lines.append(line[1:])
    return lines


def validate_bugbot_diff(diff: str) -> list[BugbotFinding]:
    findings: list[BugbotFinding] = []
    changed_files = list_changed_files(diff)
    payments_changed = any(path.startswith("src/payments/") for path in changed_files)
    payments_tests_changed = any(
        path.startswith("tests/payments/") for path in changed_files
    )

    checkout_text = "\n".join(added_lines_for_file(diff, "src/payments/checkout.ts"))

    if re.search(r"idempotencyKey:\s*token\b", checkout_text) or re.search(
        r"idempotencyKey:\s*[^,\n}]+token", checkout_text
    ):
        findings.append(
            BugbotFinding(
                rule="unsafe-idempotency-key",
                severity="blocking",
                title="Unsafe idempotency key",
                path="src/payments/checkout.ts",
                detail="Client token reused as idempotencyKey. Generate a server-side UUID instead.",
            )
        )

    if (
        payments_changed
        and re.search(r"gateway\.refund|/refunds", checkout_text)
        and not re.search(r"authorize|ownership|role|permission", checkout_text, re.I)
    ):
        findings.append(
            BugbotFinding(
                rule="refund-missing-authorization",
                severity="blocking",
                title="Refund path missing authorization",
                path="src/payments/checkout.ts",
                detail="New refund path calls gateway.refund without an authorization check.",
            )
        )

    if payments_changed and not payments_tests_changed:
        findings.append(
            BugbotFinding(
                rule="missing-payments-tests",
                severity="blocking",
                title="Missing payments tests",
                path="src/payments/",
                detail="Payment handler files changed without matching tests under tests/payments/.",
            )
        )

    return findings


def format_findings(findings: list[BugbotFinding]) -> str:
    if not findings:
        return "No Bugbot rule violations detected in the fixture diff."
    return "\n\n".join(
        f"[{finding.severity}] {finding.title}\n"
        f"  rule: {finding.rule}\n"
        f"  path: {finding.path}\n"
        f"  {finding.detail}"
        for finding in findings
    )
